[Hub Generated] Review request for Microsoft.AzureStackHCI to add version preview/2023-07-01-preview

wyatt-troia-msft opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://openapihub.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/wyatt-troia-msft-azurestackhci-Microsoft.AzureStackHCI-2023-07-01-preview...main/).
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5e91ff2d-1f5c-4e96-87c2-390e7451da37)

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [X] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [X] I have reviewed the general guidance on the spec PR review process: https://aka.ms/specprreview.
- [X] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data-plane related specifications.
- [X] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy).
- [X] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can request review from an ARM API Review board.

### ARM API changes review

- If you want for the ARM team to review this PR, you must add the `ARMReview` label. 
- The automation may automatically add the `ARMReview` label, if appropriate.  
  If this happens, proceed according to guidance given in GitHub comments also added by the automation.

### Breaking change review

If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), 
follow the process outlined in the [High-level Breaking Change Process doc](https://aka.ms/breakingchangesprocess#high-level-breaking-change-process).
      
## Getting help

- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For additional help, see https://aka.ms/azsdk/support/spectools.

wyatt-troia-msft committed:
Adds base for updating Microsoft.AzureStackHCI from version preview/2022-12-15-preview to version 2023-07-01-preview

wyatt-troia-msft committed:
Updates readme

wyatt-troia-msft committed:
Updates API version in new specs and examples

wyatt-troia committed:
Add 2023-07-01-preview

wyatt-troia-msft committed:
Update readme.md

wyatt-troia-msft committed:
Update readme.md

wyatt-troia-msft committed:
Make RP name pascal-case

wyatt-troia-msft committed:
GalleryImages use "update" in 200 response description

wyatt-troia-msft committed:
Update galleryImages.json - Add "location" header to 202 response

wyatt-troia-msft committed:
Update galleryImages.json - add location header to 202

wyatt-troia-msft committed:
Update galleryImages.json - Add type:object

wyatt-troia-msft committed:
Update marketplaceGalleryImages.json - Add 202 location header

wyatt-troia-msft committed:
Update marketplaceGalleryImages.json - pascalcase RP name

wyatt-troia-msft committed:
Update marketplaceGalleryImages.json - add type:object

wyatt-troia-msft committed:
Update networkInterfaces.json - 200 PUT description must include UPDATE

wyatt-troia-msft committed:
Update networkInterfaces.json - add 202 location header

wyatt-troia-msft committed:
Update networkInterfaces.json - add type:object

wyatt-troia-msft committed:
Update storageContainers.json - pascalcase RP name

wyatt-troia-msft committed:
Update storageContainers.json - 200 PUT description must include UPDATE

wyatt-troia-msft committed:
Update storageContainers.json - 202 location header

wyatt-troia-msft committed:
Update storageContainers.json - add type:object

wyatt-troia-msft committed:
Update virtualHardDisks.json - pascalcase RP name

wyatt-troia-msft committed:
Update virtualHardDisks.json - 202 location header

wyatt-troia-msft committed:
Update virtualHardDisks.json - 200 PUT description must include update

wyatt-troia-msft committed:
Update galleryImages.json - "x-ms-long-running-operation": true,

wyatt-troia-msft committed:
Update virtualMachineInstances.json - 200 PUT updated in description

wyatt-troia-msft committed:
Update virtualMachineInstances.json - location header 202

wyatt-troia-msft committed:
Update virtualMachineInstances.json - 	Description of 200 response code of a PUT operation MUST include term 'update'.

wyatt-troia-msft committed:
Update virtualNetworks.json - A 202 response should include an Location response header.

wyatt-troia-msft committed:
Update virtualNetworks.json - camelcase id property name

wyatt-troia-msft committed:
Update virtualNetworks.json - ID exception from camelcase

wyatt-troia-msft committed:
Update virtualNetworks.json

wyatt-troia-msft committed:
Update virtualMachineInstances.json - 202 accepted location header

wyatt-troia-msft committed:
Update virtualMachineInstances.json - add identity property to VMI update request body

wyatt-troia-msft committed:
Update virtualMachineInstances.json - move identity

wyatt-troia-msft committed:
Update virtualNetworks.json - Id camelcase attempt

wyatt-troia-msft committed:
Update virtualNetworks.json - camelcase attempt "id"

wyatt-troia-msft committed:
Update galleryImages.json - remove stray comma

wyatt-troia-msft committed:
Update marketplaceGalleryImages.json - remove stray comma

wyatt-troia-msft committed:
Update networkInterfaces.json

wyatt-troia-msft committed:
Update storageContainers.json - remove stray comma

wyatt-troia-msft committed:
Update virtualHardDisks.json

wyatt-troia-msft committed:
Update virtualMachineInstances.json - remove stray comma

wyatt-troia-msft committed:
Update virtualNetworks.json - remove stray comma

wyatt-troia committed:
Run prettier

wyatt-troia committed:
Revert to HCI

wyatt-troia-msft committed:
Update virtualHardDisks.json - type object

wyatt-troia-msft committed:
Update virtualMachineInstances.json - type object

wyatt-troia-msft committed:
Update virtualNetworks.json - type object

wyatt-troia-msft committed:
Update virtualMachineInstances.json - array id

wyatt-troia-msft committed:
Update virtualMachineInstances.json - comma

wyatt-troia-msft committed:
Update virtualMachineInstances.json - final-state-via location

wyatt-troia-msft commented:
The private specs repo PRs these changes are comprised of are listed [here](The PRs are the commit history [here](https://github.com/Altayl/azure-rest-api-specs-pr/commits/RPSaaSMaster/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-07-01-preview/virtualMachineInstances.json).).

wyatt-troia-msft committed:
Update readme.md

zizw123 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-07-01-preview/virtualMachineInstances.json:
```diff
@@ -0,0 +1,1545 @@
+          "type": "object",
+          "properties": {
+            "networkInterfaces": {
+              "type": "array",
```
>array [](http://example.com/codeflow?start=23&length=5)

use x-ms-identifiers for the object in the array: [https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-identifiers](https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-identifiers)

wyatt-troia-msft replied:
That rule says "The default identifying property is id. You don't need to add this annotataion if there is the id property in the object model." The items in this array of networkInterfaces have a unique "id" property, so we don't need to add the x-ms-identifiers annotation, right?

zizw123 replied:
Yes, in this case you are good

zizw123 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-07-01-preview/virtualMachineInstances.json:
```diff
@@ -0,0 +1,1545 @@
+                  "description": "Defines the amount of extra memory that should be reserved for a virtual machine instance at runtime, as a percentage of the total memory that the virtual machine instance is thought to need. This only applies to virtual systems with dynamic memory enabled. This property can be in the range of 5 to 2000.",
+                  "format": "int32"
+                }
+              }
```
Any min/max/default for these values?

pathuang-sia replied:
dynamicMemoryConfig is only relevant when VM does not use a static memory. We could add a min max but giving it a default value will make it always return a targetMemoryBuffer which is not relevant in static memory.

wyatt-troia-msft replied:
We are trying to stay in sync with some partner RPs (like [SCVMM ](https://github.com/Azure/azure-rest-api-specs-pr/blob/bb15ae0084afda87b7d50e3e5a0e4ac66d0e5f9d/specification/scvmm/resource-manager/Microsoft.ScVmm/preview/2023-04-01-preview/scvmm.json#L4939C16-L4939C16)) who are implementing the same contract for equivalent child virtualMachineInstance resources. They do not include a min, max, or default value. So I would prefer not to be different than they are.

wyatt-troia-msft replied:
Resolving this since @zizw123 said to just apply what was appropriate:

<img width="743" alt="image" src="https://github.com/Azure/azure-rest-api-specs/assets/84938589/c5bff601-163e-455d-996b-a4dd3e5cf018">

zizw123 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-07-01-preview/virtualMachineInstances.json:
```diff
@@ -0,0 +1,1545 @@
+              "type": "object",
+              "properties": {
+                "disablePasswordAuthentication": {
+                  "type": "boolean",
```
>boolean [](http://example.com/codeflow?start=27&length=7)

Consider enum over boolean for best practice

wyatt-troia-msft replied:
Please see [this comment](https://github.com/Azure/azure-rest-api-specs-pr/pull/13356#issuecomment-1640658444) from our RPSaaSMaster PR explaining why we would not like to do that.

wyatt-troia-msft replied:
Resolving this since @zizw123 said he didn't have additional feedback.

zizw123 commented:
> @rkmanda Here are the different LintDiff error categories we're facing:
> 
> * "Operations API must be implemented for 'Microsoft.AzureStackHCI'"
>   
>   * We're RPSaaS, so we shouldn't have to do this, right?
> * "Resource provider naming must follow the pascal case"
>   
>   * We already use "AzureStackHCI" (instead of "AzureStackHci") in production, so this would be very disruptive to change. Can we get an exception?
> * "Get endpoints for collections of resources must only have the value and nextLink properties in their model."
>   
>   * LintDiff does not seem to be correctly understanding that a path like "/{resourceUri}/providers/Microsoft.AzureStackHCI/virtualMachineInstances/default" refers to a singular resource that is always named "default," not a collection of resources.
> * "The top-level resource 'VirtualMachineInstance' does not have list by subscription operation, please add it."
>   
>   * This operation wouldn't be very useful, since there can only be one VirtualMachineInstance for each parent HybridCompute Machine resource. Do we still need to support it?
> 
> We're also facing a lot of Avocado errors. All of them seem to be complaining that API paths present in old API versions are no longer present in the new preview API version, but this is because those paths were deprecated.

The first three seem to be good, please follow this guidance to suppress these error: https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process

For the late one, can there be multiple HybridCompute machine resources per resource? If so how to get all VirtualMachineInstance?

wyatt-troia-msft committed:
Update readme.md - suppress errors

wyatt-troia-msft committed:
Update readme.md

wyatt-troia-msft committed:
Update readme.md

wyatt-troia-msft commented:
> > @rkmanda Here are the different LintDiff error categories we're facing:
> > 
> > * "Operations API must be implemented for 'Microsoft.AzureStackHCI'"
> >   
> >   * We're RPSaaS, so we shouldn't have to do this, right?
> > * "Resource provider naming must follow the pascal case"
> >   
> >   * We already use "AzureStackHCI" (instead of "AzureStackHci") in production, so this would be very disruptive to change. Can we get an exception?
> > * "Get endpoints for collections of resources must only have the value and nextLink properties in their model."
> >   
> >   * LintDiff does not seem to be correctly understanding that a path like "/{resourceUri}/providers/Microsoft.AzureStackHCI/virtualMachineInstances/default" refers to a singular resource that is always named "default," not a collection of resources.
> > * "The top-level resource 'VirtualMachineInstance' does not have list by subscription operation, please add it."
> >   
> >   * This operation wouldn't be very useful, since there can only be one VirtualMachineInstance for each parent HybridCompute Machine resource. Do we still need to support it?
> > 
> > We're also facing a lot of Avocado errors. All of them seem to be complaining that API paths present in old API versions are no longer present in the new preview API version, but this is because those paths were deprecated.
> 
> The first three seem to be good, please follow this guidance to suppress these error: https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process
> 
> For the late one, can there be multiple HybridCompute machine resources per resource? If so how to get all VirtualMachineInstance?

I've suppressed the first three errors.

There cannot be multiple HybridCompute machine resources per AzureStackHCI VirtualMachineInstance resource. It's a 1:1 relationship.

wyatt-troia-msft commented:
@zizw123 Thanks for the feedback. Ready for another review.

wyatt-troia-msft committed:
Update readme.md

wyatt-troia-msft commented:
@zizw123 asked me to suppress the final LintDiff error:

![image](https://github.com/Azure/azure-rest-api-specs/assets/84938589/9df59f37-342e-467a-9e3f-d3393cfe38e5)

zizw123 commented:
> > > @rkmanda Here are the different LintDiff error categories we're facing:
> > > 
> > > * "Operations API must be implemented for 'Microsoft.AzureStackHCI'"
> > >   
> > >   * We're RPSaaS, so we shouldn't have to do this, right?
> > > * "Resource provider naming must follow the pascal case"
> > >   
> > >   * We already use "AzureStackHCI" (instead of "AzureStackHci") in production, so this would be very disruptive to change. Can we get an exception?
> > > * "Get endpoints for collections of resources must only have the value and nextLink properties in their model."
> > >   
> > >   * LintDiff does not seem to be correctly understanding that a path like "/{resourceUri}/providers/Microsoft.AzureStackHCI/virtualMachineInstances/default" refers to a singular resource that is always named "default," not a collection of resources.
> > > * "The top-level resource 'VirtualMachineInstance' does not have list by subscription operation, please add it."
> > >   
> > >   * This operation wouldn't be very useful, since there can only be one VirtualMachineInstance for each parent HybridCompute Machine resource. Do we still need to support it?
> > > 
> > > We're also facing a lot of Avocado errors. All of them seem to be complaining that API paths present in old API versions are no longer present in the new preview API version, but this is because those paths were deprecated.
> > 
> > 
> > The first three seem to be good, please follow this guidance to suppress these error: https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process
> > For the late one, can there be multiple HybridCompute machine resources per resource? If so how to get all VirtualMachineInstance?
> 
> I've suppressed the first three errors.
> 
> There cannot be multiple HybridCompute machine resources per AzureStackHCI VirtualMachineInstance resource. It's a 1:1 relationship.

Operations API is required even for RpaaS. RPaaS provides a default implementation of the API, each service needs to document it in their own swaggers - please remove this from suppression and update accordingly

"The top-level resource 'VirtualMachineInstance' does not have list by subscription operation, please add it." If this is truly top level and not an extension resource, you'll need to implement a list by subscription even for singleton collections - please remove this from suppression and update accordingly

wyatt-troia committed:
Add operations API'

wyatt-troia committed:
Remove Operations API suppression

wyatt-troia committed:
Fix RP path

wyatt-troia committed:
Fix RP path

wyatt-troia committed:
Add operations API to gallery images

wyatt-troia committed:
Fix operations schemas

wyatt-troia committed:
Use newer operations api

wyatt-troia committed:
Remove unused schema

wyatt-troia committed:
Add ListOperations example

wyatt-troia-msft commented:
> > > > @rkmanda Here are the different LintDiff error categories we're facing:
> > > > 
> > > > * "Operations API must be implemented for 'Microsoft.AzureStackHCI'"
> > > >   
> > > >   * We're RPSaaS, so we shouldn't have to do this, right?
> > > > * "Resource provider naming must follow the pascal case"
> > > >   
> > > >   * We already use "AzureStackHCI" (instead of "AzureStackHci") in production, so this would be very disruptive to change. Can we get an exception?
> > > > * "Get endpoints for collections of resources must only have the value and nextLink properties in their model."
> > > >   
> > > >   * LintDiff does not seem to be correctly understanding that a path like "/{resourceUri}/providers/Microsoft.AzureStackHCI/virtualMachineInstances/default" refers to a singular resource that is always named "default," not a collection of resources.
> > > > * "The top-level resource 'VirtualMachineInstance' does not have list by subscription operation, please add it."
> > > >   
> > > >   * This operation wouldn't be very useful, since there can only be one VirtualMachineInstance for each parent HybridCompute Machine resource. Do we still need to support it?
> > > > 
> > > > We're also facing a lot of Avocado errors. All of them seem to be complaining that API paths present in old API versions are no longer present in the new preview API version, but this is because those paths were deprecated.
> > > 
> > > 
> > > The first three seem to be good, please follow this guidance to suppress these error: https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process
> > > For the late one, can there be multiple HybridCompute machine resources per resource? If so how to get all VirtualMachineInstance?
> > 
> > 
> > I've suppressed the first three errors.
> > There cannot be multiple HybridCompute machine resources per AzureStackHCI VirtualMachineInstance resource. It's a 1:1 relationship.
> 
> Operations API is required even for RpaaS. RPaaS provides a default implementation of the API, each service needs to document it in their own swaggers - please remove this from suppression and update accordingly
> 
> "The top-level resource 'VirtualMachineInstance' does not have list by subscription operation, please add it." If this is truly top level and not an extension resource, you'll need to implement a list by subscription even for singleton collections - please remove this from suppression and update accordingly

Ok, I removed the OperationsAPI suppression and added the Operations API.

VirtualMachineInstance is an extension resource, so are we ok to leave it as is?

wyatt-troia-msft committed:
Update galleryImages.json to use v3 common types

pathuang-sia commented:
The operations API is already carried by the partner team that we share the AzureStackHCI resource provider with: https://github.com/Azure/azure-rest-api-specs-pr/blob/RPSaaSMaster/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/stable/2023-06-01/operations.json

@rkmanda Is it still necessary for us to define the operations api here in this version?

ms-zhenhua commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-07-01-preview/virtualMachineInstances.json:
```diff
@@ -0,0 +1,1545 @@
+          },
+          "200": {
+            "description": "OK"
+          },
```
LRO-Delete should define 202 and 204 only.
[https://github.com/Azure/typespec-azure/issues/3167](https://github.com/Azure/typespec-azure/issues/3167)

ms-zhenhua commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-07-01-preview/virtualMachineInstances.json:
```diff
@@ -0,0 +1,1545 @@
+        "responses": {
+          "200": {
+            "description": "OK"
+          },
```
LRO-Delete should define 202 and 204 only.
[https://github.com/Azure/typespec-azure/issues/3167](https://github.com/Azure/typespec-azure/issues/3167)

ms-zhenhua commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-07-01-preview/virtualMachineInstances.json:
```diff
@@ -0,0 +1,1545 @@
+          "items": {
+            "$ref": "#/definitions/InstanceViewStatus"
+          },
+          "x-ms-identifiers": [],
```
>[] [](http://example.com/codeflow?start=30&length=2)

is it possible to use `code` as the identifying property?

wyatt-troia-msft replied:
@pathuang-sia Is an `InstanceViewStatus`' `code` guaranteed to be unique in the array of `InstanceViewStatuses`?

pathuang-sia replied:
compute kept it empty so we did the same to algin with it: https://github.com/Azure/azure-rest-api-specs-pr/blob/a3089b2764df11b10a8a43e2a5a33dbe4d3653ca/specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-03-01/virtualMachine.json#L1939

Is this a necessary change @ms-zhenhua? We understand it may be an improvement, but we would rather model over compute as how we did for previous changes suggested if this is not absolute necessary.

ms-zhenhua replied:
currently it's optional. But if it is not set, the elements may be duplicated which may bring unexpected errors.

ms-zhenhua commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-07-01-preview/virtualMachineInstances.json:
```diff
@@ -0,0 +1,1545 @@
+            "modelAsString": true
+          }
+        },
+        "displayStatus": {
```
>displayStatus [](http://example.com/codeflow?start=9&length=13)

can it be defined as enum?

wyatt-troia-msft replied:
@pathuang-sia What do you think? Do you know which code generates this `displayStatus`?

wyatt-troia-msft committed:
Update virtualMachineInstances.json - remove 200 from delete

wyatt-troia-msft committed:
Update virtualMachineInstances.json - remove delete 200

pathuang-sia commented:
> The operations API is already carried by the partner team that we share the AzureStackHCI resource provider with: https://github.com/Azure/azure-rest-api-specs-pr/blob/RPSaaSMaster/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/stable/2023-06-01/operations.json
> 
> @rkmanda Is it still necessary for us to define the operations api here in this version?

@ms-zhenhua Can you help answer this?

raych1 commented:
I approved go bc as it's beta version and they are intended api change.

rkmanda commented:
Please fix the Avocado and ModelValidation failures. These are blocking the merge.

wyatt-troia-msft committed:
Update DeleteVirtualMachineInstance.json - remove 200 response

wyatt-troia-msft committed:
Update DeleteGuestAgent.json - remove 200 response

wyatt-troia-msft committed:
Update GetGalleryImage.json - dummy change to retrigger CI

wyatt-troia-msft commented:
> Please fix the Avocado and ModelValidation failures. These are blocking the merge.

I have removed the 200 example responses that were causing the ModelValidation failures, but the CI pipeline no longer seems to be running the automated validations. How can we fix this?

The Avocado failures are due to intentionally-deprecated API paths.

konrad-jamrozik commented:
/azp run

konrad-jamrozik commented:
/pr RequestMerge
